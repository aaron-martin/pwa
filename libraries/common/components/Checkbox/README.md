# Checkbox
---
A simple checkbox component. It holds no state on it's own and must be controlled.

## Getting Started

```markup
import { Checkbox } from 'Library/components';

<Checkbox
  checkedIcon="X"
  uncheckedIcon="O"
  checked={someControlVariable}
/>
```

## Props

### checked

_Type_: `boolean`<br>

Defines whether the checkbox is checked or not. Used to control the checkbox. Must not be used in conjunction with the `defaultChecked` property!

###### Usage:

```
<Checkbox
  checked={this.state.checkboxIsChecked}
  ...
/>
```

### checkedIcon (required)

_Type_: `node`<br>

This is used to be shown as the checked icon.

###### Usage:

```
<Checkbox
  checkedIcon={<Checked />}
  ...
/>
```

### uncheckedIcon (required)

_Type_: `node`<br>

This is used to be shown as the unchecked icon.

###### Usage:

```
<Checkbox
  uncheckedIcon={<Unchecked />}
  ...
/>
```

### className

_Default_: `undefined`<br>
_Type_: `string`<br>

CSS class names to style the checkbox.

###### Usage:

```
<Checkbox
  className="awesome-checkbox"
  ...
/>
```

### defaultChecked

_Default_: `undefined`<br>
_Type_: `boolean`<br>

The default value for an uncontrolled checkbox. Must not be used in conjunction with the `checked` property!

###### Usage:

```
<Checkbox
  defaultChecked={false}
  ...
/>
```

### label

_Default_: `null`<br>
_Type_: `string`<br>

The label of the checkbox.

###### Usage:

```
<Checkbox
  label="I have read and agree to the privacy policy"
  ...
/>
```

### labelPosition

_Default_: `'left'`<br>
_Type_: `string`<br>

Is the label on the left or right.

###### Usage:

```
<Checkbox
  labelPosition="right"
  ...
/>
```

### name

_Default_: `undefined`<br>
_Type_: `string`<br>

If the checkbox is used in a form-tag, the name prop is used as the input-tag's name attribute.

###### Usage:

```
<Checkbox
  name="myCheckbox"
  ...
/>
```

### onCheck

_Default_: `() => {}`<br>
_Type_: `function`<br>

The callback function to be triggerd if the checkbox is clicked.
It receives a boolean indicating if the checkbox is checked.

###### Usage:

```
<Checkbox
  onCheck={checked => this.setState({ checked })}
  ...
/>
```
