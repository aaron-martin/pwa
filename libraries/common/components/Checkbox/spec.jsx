/**
 * Copyright (c) 2017, Shopgate, Inc. All rights reserved.
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React from 'react';
import { shallow } from 'enzyme';
import Checkbox from './index';

/**
 * Checked Icon
 * @returns {JSX}
 */
const Checked = () => <div />;

/**
 * Unchecked Icon
 * @returns {JSX}
 */
const Unchecked = () => <div />;

describe('<Checkbox />', () => {
  it('should render the checkbox with the label before the icon', () => {
    const wrapper = shallow(
      <Checkbox
        label={<span>Test Label Deluxe</span>}
        labelPosition="left"
        checkedIcon={<Checked />}
        uncheckedIcon={<Unchecked />}
        checked={false}
      />
    );

    const expected = (
      <div>
        <span>Test Label Deluxe</span>
        <Unchecked />
      </div>
    );

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.matchesElement(expected)).toBeTruthy();
  });

  it('should render the checkbox with the label after the icon', () => {
    const wrapper = shallow(
      <Checkbox
        label={<span>Test Label Deluxe</span>}
        labelPosition="right"
        checkedIcon={<Checked />}
        uncheckedIcon={<Unchecked />}
        checked={false}
      />
    );

    const expected = (
      <div>
        <Unchecked />
        <span>Test Label Deluxe</span>
      </div>
    );

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.matchesElement(expected)).toBeTruthy();
  });

  it('should render the unchecked icon if "checked" is false', () => {
    const wrapper = shallow(
      <Checkbox
        checked={false}
        label="Test Label Deluxe"
        checkedIcon={<Checked />}
        uncheckedIcon={<Unchecked />}
      />
    );

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(Checked).length).toBe(0);
    expect(wrapper.find(Unchecked).length).toBe(1);
  });

  it('should render the unchecked icon if "checked" is false', () => {
    const wrapper = shallow(
      <Checkbox
        checked
        label="Test Label Deluxe"
        checkedIcon={<Checked />}
        uncheckedIcon={<Unchecked />}
      />
    );

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(Checked).length).toBe(1);
    expect(wrapper.find(Unchecked).length).toBe(0);
  });

  it('should call the callback with the inverted value', () => {
    const spy = jest.fn();

    const wrapper = shallow(
      <Checkbox
        label="Test Label Deluxe"
        checkedIcon={<Checked />}
        uncheckedIcon={<Unchecked />}
        checked={false}
        onCheck={spy}
      />
    );

    wrapper.simulate('touchTap');
    expect(spy).toHaveBeenCalledWith(true);
  });

  it('should render an <input> element if a name prop is provided', () => {
    const wrapper = shallow(
      <Checkbox
        label="Test Label Deluxe"
        checkedIcon={<Checked />}
        uncheckedIcon={<Unchecked />}
        defaultChecked={false}
        name="myCheckbox"
      />
    );

    const input = wrapper.find('input');
    expect(input.length).toBe(1);
    expect(input.prop('name')).toEqual('myCheckbox');
    expect(input.prop('value')).toEqual(0);
  });

  it('should work as an uncontrolled input', () => {
    const wrapper = shallow(
      <Checkbox
        label="Test Label Deluxe"
        checkedIcon={<Checked />}
        uncheckedIcon={<Unchecked />}
        defaultChecked={false}
      />
    );

    wrapper.simulate('touchTap');
    expect(wrapper.state('checked')).toBe(true);
  });
});
