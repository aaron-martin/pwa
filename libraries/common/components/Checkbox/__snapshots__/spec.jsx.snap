// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Checkbox /> should render the checkbox with the label after the icon 1`] = `
<div
  onTouchTap={[Function]}
>
  <Unchecked />
  <span>
    Test Label Deluxe
  </span>
</div>
`;

exports[`<Checkbox /> should render the checkbox with the label before the icon 1`] = `
<div
  onTouchTap={[Function]}
>
  <span>
    Test Label Deluxe
  </span>
  <Unchecked />
</div>
`;

exports[`<Checkbox /> should render the unchecked icon if "checked" is false 1`] = `
<div
  onTouchTap={[Function]}
>
  Test Label Deluxe
  <Unchecked />
</div>
`;

exports[`<Checkbox /> should render the unchecked icon if "checked" is false 2`] = `
<div
  onTouchTap={[Function]}
>
  Test Label Deluxe
  <Checked />
</div>
`;
